### PR TITLE
Timeout increase for consecutive rounds [ECR-1281]

### PR DESCRIPTION
**Pull request by dvush**
Here the size of the timeout increase is introduced as a parameter.

### CODE REVIEW

**Comment by ilammy** [comment]
I believe this definitely warrants an entry in changelog.

File: exonum/src/blockchain/config.rs [hunk 1]
```diff
@@ -89,16 +89,19 @@ pub struct StoredConfiguration {
...
     /// been committed during this period or not.
+    /// Each consecutive round will be longer then previous by round_timeout_increase.
     ///
     /// Note that rounds in Exonum
     /// do not have a defined end time. Nodes in a new round can
     /// continue to vote for proposals and process messages related to previous
     /// rounds.
-    pub round_timeout: Milliseconds,
+    pub first_round_timeout: Milliseconds,
+    /// Amount of time that is added to round timeout as round number increase.
```
Comment (vitvakatu) [inline exonum/src/blockchain/config.rs]: I think `Amount of time that is added to the round timeout each round` is better.

File: exonum/src/node/mod.rs [hunk 2]
```diff
@@ -676,7 +681,9 @@ impl NodeHandler {
     /// Returns start time of the requested round.
     pub fn round_start_time(&self, round: Round) -> SystemTime {
         let previous_round: u64 = round.previous().into();
-        let ms = previous_round * self.round_timeout();
+        let ms = previous_round * self.first_round_timeout()
+            + (previous_round * previous_round.saturating_sub(1)) / 2
```
Comment (vitvakatu) [inline exonum/src/node/mod.rs]: Could you please clarify this formulae?

File: exonum/src/blockchain/config.rs [hunk 3]
```diff
@@ -89,16 +89,19 @@ pub struct StoredConfiguration {
...
+    pub round_timeout_increase: Milliseconds,
```
Comment (vitvakatu) [inline exonum/src/blockchain/config.rs]: I don't think it should be constant value

File: exonum/src/blockchain/config.rs [hunk 4]
```diff
@@ -89,16 +89,19 @@ pub struct StoredConfiguration {
...
+    /// time when second round starts, regardless of whether a new block has
     /// been committed during this period or not.
+    /// Each consecutive round will be longer then previous by round_timeout_increase.
     ///
     /// Note that rounds in Exonum
     /// do not have a defined end time. Nodes in a new round can
     /// continue to vote for proposals and process messages related to previous
     /// rounds.
-    pub round_timeout: Milliseconds,
+    pub first_round_timeout: Milliseconds,
```
Comment (vitvakatu) [inline exonum/src/blockchain/config.rs]: Why not `initial_round_timeout`?

File: exonum/src/node/mod.rs (hunk 2 above)
Comment (dvush) [reply exonum/src/node/mod.rs]: It is the sum of the arithmetic progression. For round r cumulative increase in round start time due to increasing round timeout will be (r-1)(r-2)/2 * round_timeout_increase. I'll clarify it in the comment.

File: exonum/src/blockchain/config.rs (hunk 4 above)
Comment (dvush) [reply exonum/src/blockchain/config.rs]: It is equivalent but I like first_* since it is more concrete.

**Comment by dvush** [comment]
@vitvakatu 
>dismiss previous review
Why is it still blocked then?

File: exonum/src/sandbox/consensus/round_details.rs [hunk 5]
```diff
@@ -1335,8 +1335,8 @@ fn handle_precommit_positive_scenario_commit_with_queued_precommit() {
     let sandbox = timestamping_sandbox();
     let sandbox_state = SandboxState::new();
 
-    let block_1_delay = 2 * sandbox.round_timeout() + PROPOSE_TIMEOUT + 1;
-    let block_2_delay = 2 * sandbox.round_timeout() + 2 * PROPOSE_TIMEOUT + 1;
+    let block_1_delay = 2 * sandbox.current_round_timeout() + PROPOSE_TIMEOUT + 1;
+    let block_2_delay = 2 * sandbox.current_round_timeout() + 2 * PROPOSE_TIMEOUT + 1;
```
Comment (ilammy) [inline exonum/src/sandbox/consensus/round_details.rs]: I think if round timeout is not fixed then we can't do `2 * sandbox.current_round_timeout()` anymore and have to be more elaborate. It may work with current configuration values, but may suddenly break if someone tweaks them.

Comment (dvush) [reply exonum/src/sandbox/consensus/round_details.rs]: I thought about it, but this parameter is not used by block builder anyway, should be revisited.

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,11 +16,17 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 
 - Config parameter `external_address` is now a required value. (#826)
 
+- Config parameter `round_timeout` is removed.
```
Comment (vitvakatu) [inline CHANGELOG.md]: We prefer to use present perfect tense in changelog.

File: CHANGELOG.md [hunk 7]
```diff
@@ -16,11 +16,17 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
...
+  `first_round_timeout` and `round_timeout_increase` should be used instead. (#848)
```
Comment (vitvakatu) [inline CHANGELOG.md]: I suggest to elaborate on it a little bit, just to make sure user will understand how to use new feature. And haven't you checked all the documentation? We need to update it too (perhaps in a different PRs)

File: exonum/src/blockchain/config.rs [hunk 8]
```diff
@@ -89,16 +89,19 @@ pub struct StoredConfiguration {
...
+    /// Amount of time that is added to the round timeout each round
```
Comment (DarkEld3r) [inline exonum/src/blockchain/config.rs]: Please add a dot.

File: exonum/src/blockchain/config.rs [hunk 9]
```diff
@@ -89,16 +89,19 @@ pub struct StoredConfiguration {
...
+    pub round_timeout_increase: Milliseconds,
```
Comment (DarkEld3r) [inline exonum/src/blockchain/config.rs]: Should this value be configurable? Can we simply use the percentage of `first_round_timeout`?

File: exonum/src/blockchain/config.rs [hunk 10]
```diff
@@ -138,11 +141,11 @@ impl ConsensusConfig {
         const MIN_TXS_BLOCK_LIMIT: u32 = 100;
         const MAX_TXS_BLOCK_LIMIT: u32 = 10_000;
 
-        if self.round_timeout <= 2 * self.max_propose_timeout {
+        if self.first_round_timeout <= 2 * self.max_propose_timeout {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/config.rs]: I suppose some validation of `round_timeout_increase` should be added too. Do you intentionally allow zero value?

File: CHANGELOG.md [hunk 11]
```diff
@@ -16,16 +16,20 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
...
 - Config parameter `external_address` is now a required value. (#826)
 
-- Config parameter `round_timeout` is removed.
-  `first_round_timeout` and `round_timeout_increase` should be used instead. (#848)
+- Config parameter `round_timeout` has been renamed to `first_round_timeout`.
+  Now timeout for round r is
+  `first_round_timeout + (r-1)*round_timeout_increase`
+   where `round_timeout_increase` is determined as a some percentage of
+  `first_round_timeout`. Value of this percentage is defined in constant inside
+  `ConsensusConfig` (#848)
```
Comment (ilammy) [inline CHANGELOG.md]: Some last-minute nit: I'd reword this changelog entry to explicitly reference the `TIMEOUT_LINEAR_INCREASE_PERCENT` constant instead of 'some percentage' (it is public after all). We can also note its current value (10%) so that developers reading the changelog have an idea of its magnitude.

**Comment by nemat-rakhmatov** [comment]
Hi all,

From the above conversations I can not quite get why this feature was implemented? Could someone shed the light into it? What were the requirements to this change?

Thanks.

**Comment by DarkEld3r** [comment]
@nemat-rakhmatov 
Hello,

In Exonum we suppose that the network is partially synchronous (see https://exonum.com/blog/10-13-17-consensus-message-delivery-and-time/ for details): each message is delivered faster than unknown finite time T. As we don’t know this time T we should increase round timeout to have a stable BFT consensus. In future we are planning to implement adaptive round timeout adjustment (see https://exonum.com/blog/adjustable-timeouts-and-empty-blocks-2/).

**Comment by nemat-rakhmatov** [comment]
Thanks for explanation (and links). Things getting more or less clear now.
